dev/src/document.ts

```diff
@@ -23,9 +23,90 @@ import is from 'is';
 import {FieldPath} from './path';
 import {FieldTransform} from './field-value';
 import {Timestamp} from './timestamp';
-import {validatePkg} from './validate';
 import {isPlainObject} from './serializer';
 
+/**
+ * Returns a builder for DocumentSnapshot and QueryDocumentSnapshot instances.
+ * Invoke `.build()' to assemble the final snapshot.
+ *
+ * @private
+ * @class
+ */
+class DocumentSnapshotBuilder {
+  ref;
+  fieldsProto;
+  readTime;
+  createTime;
+  updateTime;
+  /**
+   * @private
+   * @hideconstructor
+   *
+   * @param {DocumentSnapshot=} snapshot An optional snapshot to base this
+   * builder on.
+   */
+  constructor(snapshot?) {
+    snapshot = snapshot || {};
+
+    /**
+     * The reference to the document.
+     *
+     * @type {DocumentReference}
+     */
+    this.ref = snapshot._ref;
+
+    /**
+     * The fields of the Firestore `Document` Protobuf backing this document.
+     *
+     * @type {object}
+     */
+    this.fieldsProto = snapshot._fieldsProto;
+
+    /**
+     * The time when this document was read.
+     *
+     * @type {Timestamp}
+     */
+    this.readTime = snapshot._readTime;
+
+    /**
+     * The time when this document was created.
+     *
+     * @type {Timestamp}
+     */
+    this.createTime = snapshot._createTime;
+
+    /**
+     * The time when this document was last updated.
+     *
+     * @type {Timestamp}
+     */
+    this.updateTime = snapshot._updateTime;
+  }
+
+  /**
+   * Builds the DocumentSnapshot.
+   *
+   * @private
+   * @returns {QueryDocumentSnapshot|DocumentSnapshot} Returns either a
```

Comment:
In my conversion, I removed the JSDoc types from the comments for non public methods. It's still needed for public methods to generate the docs.

Reply:
I kind of want to keep them around for now.  We can always come back and clean this kind of stuff up.